proposals/0148-stake-program-move-instructions.md

```diff
@@ -0,0 +1,201 @@
+`amount`, but this would require changes to `Split` to enable the first leg of
+the proposed flow.
+* Back out the changes introduced by requiring rent-exempt `Split` destinations.
+This is undesirable because that restriction was added for very good reason: an
+effectively unbounded amount of stake could be instantly deactivated through
+repeated splitting.
+
+## New Terminology
+
+`MoveStake` and `MoveLamports`, two new stake program instructions.
+
+## Detailed Design
+
+For clarity of terminology inside this specification:
+
+* An "active" stake is in a `Stake` state with nonzero delegation, 100% of which
+is effective stake. There is no activating or deactivating stake.
+* An "inactive" stake is in an `Initialized` or `Stake` state. There is no
+effective, activating, or deactivating stake.
+
+### `MoveStake`
+
+`MoveStake` requires 5 accounts:
+
+0. Source stake account: Writable, owned by the stake program
+1. Destination stake account: Writable, owned by the stake program
+2. Clock: Read-only, with the fixed address
+`SysvarC1ock11111111111111111111111111111111`
+3. Stake history: Read-only, with the fixed address
+`SysvarStakeHistory1111111111111111111111111`
+4. Stake account authority: Read-only signer
+
+`MoveStake` instruction data is 12 bytes, containing:
+
+* `0x10 0x00 0x00 0x00`, a fixed-value four-byte little-endian unsigned integer
+acting as the instruction discriminator
+* `amount`, an unaligned eight-byte little-endian unsigned integer indicating
+the quantity of lamports to move
+
+`MoveStake` aborts the transaction when:
+
+* `amount` is 0
+* Source or destination are not writable
+* Source or destination are not owned by the stake program
+* Source and destination have the same address
+* Source and destination do not have identical `Authorized`
+* If `Lockup` is in force, source and destination do not have identical `Lockup`
+* The stake account authority is not the `Staker` on both accounts
+* The stake account authority is not a signer
+* Source data length is not equal to the current version of `StakeState`
+* Destination data length is not equal to the current version of `StakeState`
+* Source is not active
+* Destination is neither active nor inactive
+* If destination is active, source and destination are not delegated to the same
+vote account
+* Moving `amount` stake would leave source with a nonzero amount of stake less
+than the minimum delegation
+* Moving `amount` stake would fail to bring destination up to minimum delegation
+
+If all of these conditions hold, then:
+
+* Delegation and lamports on source are debited `amount`
+* Delegation and lamports on destination are credited `amount`
+* If `amount` constitutes the full delegation on the source, source is reset to
+an `Initialized` state
+* If destination is inactive, destination becomes active with the same `Stake`
+as source, aside from delegation amount
+* `credits_observed` must be updated on the destination according to the same
+rules as `Merge`
```

Comment:
How does this affect activation?

Reply:
it doesnt. neither instruction _ever_ activates or deactivates stake at all, their only purpose is to move value around inside a authorized/lockup bubble without changing its state. _accounts_ may become active or inactive based on `MoveStake`, but the total delegation amounts never change:
* active stake stays active and moves with its underlying lamports
* lamports in an inactive account never become active just because stake is moved into it
* accounts in any activating or deactivating state, including partially active/inactive accounts, are rejected by `MoveStake` for simplicity
* moving stake from an active account to another active account requires them to be delegated to the same vote account

Reply:
Got it great!